### ISSUE TITLE
how to extract i3d features for gtea dataset

### ISSUE THREAD

**Issue by Mythili-kannan** [Thinksky5124/SVTAS#15]
There is no file called extract_features.py under tools/extract , can you please help me here to extract features for gtea dataset from video

**Comment by VSSREEKANTH** [comment]
Hi Mythili, you can extract features using the following code

import cv2
import numpy as np
import tensorflow as tf
from tensorflow.keras.applications import InceptionV3

# Load the I3D model (pre-trained on Kinetics)
def load_i3d_model():
    base_model = InceptionV3(include_top=False, weights='imagenet', pooling='avg')
    return base_model

# Preprocess the video frames for I3D
def preprocess_frame(frame):
    frame = cv2.resize(frame, (299, 299))  # Resize to 299x299 (input size for InceptionV3)
    frame = np.expand_dims(frame, axis=0)  # Add batch dimension
    frame = frame / 255.0  # Normalize to [0, 1]
    return frame

# Extract features from the video using I3D
def extract_features_from_video(video_path, model):
    cap = cv2.VideoCapture(video_path)

    if not cap.isOpened():
        print("Error: Could not open video.")
        return None

    features = []

    while True:
        ret, frame = cap.read()
        if not ret:
            break  # Break if no frames are left

        # Preprocess the frame
        frame_preprocessed = preprocess_frame(frame)

        # Extract features
        feature_vector = model.predict(frame_preprocessed)
        features.append(feature_vector)

    cap.release()  # Release the video capture object

    # Convert features list to numpy array and reshape
    features_array = np.array(features).reshape(-1, 2048)  # 2048 is the output size for InceptionV3
    np.save(output_file, features_array)
    return features_array

def save_features_to_npy(features_array, output_file):
    np.save(output_file, np.transpose(features_array))
    

# Usage




#===============================================
import os
folder_path = '**provide_input_video_directory path** '  # Change this to your folder path
for filename in os.listdir(folder_path):
        # Check if it's a file (and not a directory)
        file_path = os.path.join(folder_path, filename)
        if os.path.isfile(file_path):
            # Split the filename and extension, keep only the filename
            file_without_extension = os.path.splitext(filename)[0]
            video_path = '**input video directory path**'+filename  # Change to your video file path
            output_file = 'directory path to save features'+file_without_extension+'.npy'

            # Load the model
            model = load_i3d_model()

            # Extract features from the video
            features_array = extract_features_from_video(video_path, model)
            print("shape of features======>", np.shape(features_array))

            if features_array is not None:
                # Save the features to a .npy file
                save_features_to_npy(features_array, output_file)
